test/s3/line_stream_test.exs

```diff
@@ -0,0 +1,51 @@
+defmodule S3.LineStreamTest do
+  use ExUnit.Case, async: true
+
+  alias S3.LineStream
+
+  describe "lines/3" do
+    test "returns content line-by-line" do
+      stream = Stream.map(["abc\nde", "fg\nhij\n"], & &1)
+
+      list =
+        stream
+        |> LineStream.lines()
+        |> Enum.to_list()
+
+      assert list == ["abc", "defg", "hij"]
+    end
+
+    test "handles \r\n newline characters" do
+      stream = Stream.map(["abc\r\nde", "fg\r\nhij\r\n"], & &1)
+
+      list =
+        stream
+        |> LineStream.lines()
+        |> Enum.to_list()
+
+      assert list == ["abc", "defg", "hij"]
+    end
+
+    test "handles no end of file newline" do
+      stream = Stream.map(["abc\r\nde", "fg\r\nhij"], & &1)
+
+      list =
+        stream
+        |> LineStream.lines()
+        |> Enum.to_list()
+
+      assert list == ["abc", "defg", "hij"]
+    end
+
+    test "handles multiple lines per chunk" do
+      stream = Stream.map(["abc\r\ndef\r\ngh", "i\r\njk\r\nlm\r\nno\r\n"], & &1)
+
+      list =
+        stream
+        |> LineStream.lines()
+        |> Enum.to_list()
+
+      assert list == ["abc", "def", "ghi", "jk", "lm", "no"]
+    end
```

Comment:
Two things:
1) two tests which might be valuable: a chunk which ends with a newline, and a chunk with no newlines.
2) You might look into `stream_data`: a property test would be a great fit here.

Reply:
Those additional tests are both great and revealed a bug (which maybe you saw coming). I also threw in one for a final chunk which ends up being just a newline character.

Using StreamData is a great idea. If it sounds good to you I split it out into a [separate ticket](https://app.asana.com/0/1199957631529739/1200086228860142) in order to take some time to understand how best to use it.